packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1459 @@
+        address indexed delegator,
+        uint256 tokenId,
+        uint128 kroShares,
+        uint128 kghShares
+    );
+
+    /**
+     * @notice Emitted when KGHs undelegation is initiated in batch.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param tokenIds  Array of token Ids of the KGHs.
+     * @param kroShares The amount of KRO shares to undelegate.
+     * @param kghShares The amount of KGH shares to undelegate.
+     */
+    event KghBatchUndelegationInitiated(
+        address indexed validator,
+        address indexed delegator,
+        uint256[] tokenIds,
+        uint128 kroShares,
+        uint128 kghShares
+    );
+
+    /**
+     * @notice Emitted when KRO undelegation is finalized.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param amount    The amount of KRO undelegated.
+     */
+    event KroUndelegationFinalized(
+        address indexed validator,
+        address indexed delegator,
+        uint128 amount
+    );
+
+    /**
+     * @notice Emitted when KGH undelegation is finalized.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param amount    The amount of KRO undelegated during the KGH undelegation.
+     */
+    event KghUndelegationFinalized(
+        address indexed validator,
+        address indexed delegator,
+        uint128 amount
+    );
+
+    /**
+     * @notice Emitted when pending challenge reward for challenge winner is distributed.
+     *
+     * @param recipient Address of the reward recipient.
+     * @param amount    The amount of challenge reward.
+     */
+    event ChallengeRewardDistributed(address recipient, uint128 amount);
+
+    /**
+     * @notice Emitted when the output reward is distributed.
+     *
+     * @param recipient Address of the reward recipient.
+     * @param amount    The amount of reward.
+     */
+    event RewardDistributed(address recipient, uint128 amount);
+
+    /**
+     * @notice Emitted when the validator is slashed.
+     *
+     * @param loser  Address of the loser at the challenge.
+     * @param winner Address of the winner at the challenge.
+     * @param amount The amount of KRO slashed.
+     */
+    event Slashed(address indexed loser, address indexed winner, uint128 amount);
+
+    /**
+     * @notice Modifier to check if the caller is the Colosseum contract.
+     */
+    modifier onlyColosseum() {
+        require(
+            msg.sender == L2_ORACLE.COLOSSEUM(),
+            "AssetManager: Only Colosseum can call this function"
+        );
+        _;
+    }
+
+    /**
+     * @notice Modifier to check if the vault is active.
+     */
+    modifier checkIsActive(address validator) {
+        require(
+            msg.sender == validator || _vault[validator].asset.validatorKro >= MIN_REGISTER_AMOUNT,
```

Comment:
The `msg.sender` should be validator here?

Reply:
Cause the validator can also delegate self, so in the case where `msg.sender` is the validator itself, we should not check if the `validatorKro` is more than `MIN_REGISTER_AMOUNT`.

Reply:
Ah, validator also delegates using `delegate` function. I see. TY!